swap-feed/src/kucoin.rs

```diff
@@ -0,0 +1,424 @@
+                    "KuCoin websocket connection failed, retrying in {}ms. Error {:#}",
+                    next.as_millis(),
+                    error
+                );
+            },
+        )
+        .await;
+
+        match result {
+            Err(e) => {
+                tracing::warn!("Rate updates incurred an unrecoverable error: {:#}", e);
+
+                // in case the retries fail permanently, let the subscribers know
+                price_update.send(Err(Error::PermanentFailure(e.into())))
+            }
+            Ok(never) => match never {},
+        }
+    });
+
+    Ok(PriceUpdates {
+        inner: price_update_receiver,
+    })
+}
+
+#[derive(Clone, Debug)]
+pub struct PriceUpdates {
+    inner: watch::Receiver<PriceUpdate>,
+}
+
+impl PriceUpdates {
+    pub async fn wait_for_next_update(&mut self) -> Result<PriceUpdate> {
+        self.inner.changed().await?;
+
+        Ok(self.inner.borrow().clone())
+    }
+
+    pub fn latest_update(&mut self) -> PriceUpdate {
+        self.inner.borrow().clone()
+    }
+}
+
+#[derive(Clone, Debug, thiserror::Error)]
+pub enum Error {
+    #[error("Rate is not yet available")]
+    NotYetAvailable,
+    #[error("Permanently failed to retrieve rate from KuCoin")]
+    PermanentFailure(Arc<anyhow::Error>),
+}
+
+type PriceUpdate = Result<wire::PriceUpdate, Error>;
+
+/// Maps a [`connection::Error`] to a backoff error, effectively defining our
+/// retry strategy.
+fn to_backoff(e: connection::Error) -> backoff::Error<anyhow::Error> {
+    use backoff::Error::*;
+
+    match e {
+        // Connection closures and websocket errors will be retried
+        connection::Error::ConnectionClosed => backoff::Error::transient(anyhow::Error::from(e)),
+        connection::Error::WebSocket(_) => backoff::Error::transient(anyhow::Error::from(e)),
+
+        // Failures while parsing a message are permanent because they most likely present a
+        // programmer error
+        connection::Error::Parse(_) | connection::Error::NoWebsocketServers => {
+            Permanent(anyhow::Error::from(e))
+        }
+    }
+}
+
+/// KuCoin websocket connection module.
+///
+/// Responsible for getting the token and address for the KuCoin websocket API,
+/// then establishing a connection to it and transforming the received websocket
+/// frames into a stream of rate updates. The connection may fail in which case
+/// it is simply terminated and the stream ends.
+mod connection {
+    use super::*;
+    use futures::stream::BoxStream;
+    use std::sync::Arc;
+    use std::time::Duration;
+    use tokio::sync::Mutex;
+    use tokio_tungstenite::tungstenite;
+
+    pub async fn new(
+        rest_url: Url,
+        client: reqwest::Client,
+    ) -> Result<BoxStream<'static, Result<wire::PriceUpdate, Error>>> {
+        let auth: wire::BulletPublicResponse = client
+            .post(rest_url)
+            .send()
+            .await
+            .context("Failed to call the KuCoin REST API")?
```

Comment:
add some more context here as to what went wrong (e.g "Failed to call KuCoin REST API to acquire auth token and websocket servers")